File: algorithms/2_medium/R/03310_remove_methods_from_project/linked_forward_star.c
struct entity {
    int v;
    int next;
};

__attribute__((no_sanitize("all")))
static inline struct entity* init_adj(
    const int n,
    int** edges, const int rows, const int* cols_data
) {
    struct entity* ptr = malloc(((size_t)n + (size_t)rows) * sizeof(*ptr));
    if (ptr == NULL) {
        fprintf(stderr, "alloc fail");
        exit(1);
    }

    for (int i = 0; i < n; i += 1) {
        ptr[i].v = -1;
        ptr[i].next = -1;
    }

    for (int i = 0; i < rows; i += 1) {
        if (cols_data[i] != 2) {
            fprintf(stderr, "invalid input");
            exit(1);
        }
        const int u = edges[i][0];
        const int v = edges[i][1];
        if (u < 0 || v < 0 || u >= n || v >= n) {
            fprintf(stderr, "invalid input");
            exit(1);
        }

        const int next = ptr[u].next;
        ptr[n + i].v = v;
        ptr[n + i].next = next;
        ptr[u].next = n + i;
    }

    return ptr;
}

__attribute__((no_sanitize("all")))
static inline bool* init_is_sus(
    const int n, const int k,
    const struct entity *restrict adj
) { 
    int *restrict stack = malloc((size_t)n * sizeof(*stack));
    bool *restrict is_sus = calloc((size_t)n, sizeof(*is_sus));
    if (is_sus == NULL || stack == NULL) {
        fprintf(stderr, "alloc fail");
        exit(1);
    }

    stack[0] = k;
    size_t stack_len = 1;
    is_sus[k] = true;

    while (stack_len > 0) {
        stack_len -= 1;
        const int u = stack[stack_len];

        int idx = adj[u].next;
        while (idx != -1) {
            const int v = adj[idx].v;
            if (!is_sus[v])  {
                stack[stack_len] = v;
                stack_len += 1;
                is_sus[v] = true;
            }
            idx = adj[idx].next;
        }
    }

    free(stack);
    return is_sus;
}
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
__attribute__((no_sanitize("all")))
int* remainingMethods(
    int n, int k,
    int** invocations, const int rows, const int* cols_data,
    int* return_len
) {
    if (n < 1 || n > 100000 || k < 0 || k >= n || rows > 200000 ||
        rows < 0 || (rows > 0 && (invocations == NULL || cols_data == NULL))
    ) {
        fprintf(stderr, "invalid input");
        exit(1);
    }

    const struct entity *restrict adj = init_adj(n, invocations, rows, cols_data);
    const bool *restrict is_sus = init_is_sus(n, k, adj);

    int* res = malloc((size_t)n * sizeof(*res));
    if (res == NULL) {
        fprintf(stderr, "alloc fail");
        exit(1);
    }

    for (int node = 0; node < n; node += 1) {
        if (!is_sus[node]) {
            int idx = adj[node].next;
            while (idx != -1) {
                const int v = adj[idx].v;
                if (is_sus[v]) {
                    for (int i = 0; i < n; i += 1) {
                        res[i] = i;
                    }
                    *return_len = n;

                    goto cleanup;
                }
                idx = adj[idx].next;
            }
        }
    }

    int res_len = 0;
    for (int i = 0; i < n; i += 1) {
        if (!is_sus[i]) {
            res[res_len] = i;
            res_len += 1;
        }
    }
    *return_len = res_len;

cleanup:
    free(adj);
    free(is_sus);
    return res;
}

File: algorithms/2_medium/R/03310_remove_methods_from_project/linked_forward_star.rs
struct Entity {
    v: i32,
    next: i32,
}

impl Solution {
    #[inline(always)]
    fn init_adj(edges: Vec<Vec<i32>>, n: i32) -> Vec<Entity> {
        let mut adj: Vec<Entity> = Vec::with_capacity(edges.len() + n as usize);
        for i in 0..n {
            adj.push(Entity { v: -1, next: -1 });
        }

        for e in edges {
            assert!(e.len() == 2);
            let (u, v) = (e[0], e[1]);
            assert!(u >= 0 && u < n && v >= 0 && v < n);
            let next = adj[u as usize].next;
            adj[u as usize].next = adj.len() as i32;
            adj.push(Entity { v: v, next: next });
        }

        adj
    }

    #[inline(always)]
    fn init_is_sus(n: i32, k: i32, adj: &[Entity]) -> Vec<bool> {
        let mut is_sus = vec![false; n as usize];
        let mut stack: Vec<i32> = Vec::with_capacity(n as usize);
        stack.push(k);
        is_sus[k as usize] = true;
        while let Some(u) = stack.pop() {
            let mut idx = adj[u as usize].next;
            while idx != -1 {
                let v = adj[idx as usize].v;
                if !is_sus[v as usize]  {
                    stack.push(v);
                    is_sus[v as usize] = true;
                }
                idx = adj[idx as usize].next;
            }
        }

        is_sus
    }

    pub fn remaining_methods(n: i32, k: i32, invocations: Vec<Vec<i32>>) -> Vec<i32> {
        assert!(n >= 1 && n <= 100_000 && k >= 0 && k < n && invocations.len() <= 200_000);
        let adj = Self::init_adj(invocations, n);
        let is_sus = Self::init_is_sus(n, k, &adj);

        for node in 0..n {
            if !is_sus[node as usize] {
                let mut idx = adj[node as usize].next;
                while idx != -1 {
                    let v = adj[idx as usize].v;
                    if is_sus[v as usize]  {
                        return (0..n).into_iter().collect::<Vec<i32>>();
                    }
                    idx = adj[idx as usize].next;
                }
            }
        }

        (0..n).into_iter().filter(|&node| !is_sus[node as usize]).collect::<Vec<i32>>()
    }
}
